src/mame/f256/f256.cpp

```diff
@@ -0,0 +1,2011 @@
+    //logerror("2D Fill DMA: DEST: %X, W: %X, H: %X\n", dest_addr, width_2D, height_2D);
+    for (int y = 0; y < height_2D; y++)
+    {
+        for (int x = 0; x < width_2D; x++)
+        {
+            m_ram->write(dest_addr + x + y * dest_stride, fill_byte);
+        }
+    }
+}
+void f256_state::performLinearFillDMA()
+{
+    uint8_t fill_byte = m_iopage[0]->read(0xDF01 - 0xC000);
+    uint32_t dest_addr = ((m_iopage[0]->read(0xDF0A) & 0x7) << 16) + (m_iopage[0]->read(0xDF09) << 8) + m_iopage[0]->read(0xDF08);
+    uint32_t count = ((m_iopage[0]->read(0xDF0E) & 0x7) << 16) + (m_iopage[0]->read(0xDF0D) << 8) + m_iopage[0]->read(0xDF0C);
+    //logerror("Linear Fill DMA DEST: %X, LEN: %X\n", dest_addr, count);
+    memset(m_ram->pointer() + dest_addr, fill_byte, count);
+}
+void f256_state::perform2DDMA()
+{
+    uint32_t src_addr = ((m_iopage[0]->read(0xDF06) & 0x7) << 16) + (m_iopage[0]->read(0xDF05) << 8) + m_iopage[0]->read(0xDF04);
+    uint32_t dest_addr = ((m_iopage[0]->read(0xDF0A) & 0x7) << 16) + (m_iopage[0]->read(0xDF09) << 8) + m_iopage[0]->read(0xDF08);
+    uint16_t width_2D = (m_iopage[0]->read(0xDF0D - 0xC000) << 8) + m_iopage[0]->read(0xDF0C - 0xC000);
+    uint16_t height_2D = (m_iopage[0]->read(0xDF0F - 0xC000) << 8) + m_iopage[0]->read(0xDF0E - 0xC000);
+    uint16_t src_stride = (m_iopage[0]->read(0xDF11 - 0xC000) << 8) + m_iopage[0]->read(0xDF10 - 0xC000);
+    uint16_t dest_stride = (m_iopage[0]->read(0xDF13 - 0xC000) << 8) + m_iopage[0]->read(0xDF12 - 0xC000);
+    //logerror("2D Copy DMA, SRC: %X, DEST: %X, W: %X H: %X, SRC_STR: %X, DEST_STR: %X\n", src_addr, dest_addr,
+    //    width_2D, height_2D, src_stride, dest_stride);
+    for (int y = 0; y < height_2D; y++)
+    {
+        for (int x = 0; x < width_2D; x++)
+        {
+            uint8_t src_byte = m_ram->read(src_addr + x + y * src_stride);
+            m_ram->write(dest_addr + x + y * dest_stride, src_byte);
+        }
+    }
+}
+void f256_state::performLinearDMA()
+{
+    uint32_t src_addr = ((m_iopage[0]->read(0xDF06) & 0x7) << 16) + (m_iopage[0]->read(0xDF05) << 8) + m_iopage[0]->read(0xDF04);
+    uint32_t dest_addr = ((m_iopage[0]->read(0xDF0A) & 0x7) << 16) + (m_iopage[0]->read(0xDF09) << 8) + m_iopage[0]->read(0xDF08);
+    uint32_t count = ((m_iopage[0]->read(0xDF0E) & 0x7) << 16) + (m_iopage[0]->read(0xDF0D) << 8) + m_iopage[0]->read(0xDF0C);
+    //logerror("Linear Copy DMA SRC: %X, DEST: %X, LEN: %X\n", src_addr, dest_addr, count);
+    memcpy(m_ram->pointer() + dest_addr, m_ram->pointer() + src_addr, count);
+}
+
+
+//-------------------------------------------------
+//  device_start
+//-------------------------------------------------
+void f256_state::device_start()
+{
+	driver_device::device_start();
+    reset_mmu();
+    // TODO: Copy the font from file to IO Page 1
+    //memcpy(m_iopage[1], m_font, 0x800);
+    for (int i=0;i<0x800;i++)
+    {
+        m_iopage[1]->write(i, m_font->as_u8(i));
```

Comment:
Is this really what happens, or should the font ROM be decoded where `m_iopage[1]` is currently being decoded?

Reply:
The machine itself loads the font when the machine is booted.  The font area is R/W, so users can modify it.  I could not figure out a more elegant want to copy the font into the iopage area, hence the commented out memcpy.

Reply:
The font is loaded at boot time.  Users can modify the font at run time.